### ISSUE TITLE
Expose TSV headers for event validation

### ISSUE THREAD

**Issue by happy5214** [bids-standard/legacy-validator#736]
#537 will require access to the column indices of `HED` and related HED headers (as defined in the sidecar) in the `_events.tsv` file. It does not appear that information is currently exposed, though it is collected in the TSV validation.

**Comment by sappelhoff** [comment]
do you mean the row values of the `HED` column in the `_events.tsv` file? Could you please clarify what you mean by 

> column indices of HED

**Comment by happy5214** [comment]
How do I determine which columns have HED tags? More generally, how do I map any optional column header (such as `HED`) to a subscript in an array formed by splitting a row by tabs?

**Comment by sappelhoff** [comment]
All of the following is valid for `events.tsv` files

1. you can expect the `HED` column to have HED tags (that's the easiest)
2. if other columns contain HED tags, this MUST be declared in an accompanying `events.json` file such that the dictionary describing the column always contains an `HED` key-value-pair

such as in this example:
```
{
    "mycodes": {
        "HED": {
            "Fixation": "Event/Category/Experimental stimulus, Event/Label/CrossFix, Event/Description/A cross appears at screen center to serve as a fixation point, Sensory presentation/Visual, Item/Object/2D Shape/Cross, Attribute/Visual/Fixation point, Attribute/Visual/Rendering type/Screen, Attribute/Location/Screen/Center",
            "Target": "Event/Label/Target image, Event/Description/A white airplane as the RSVP target superimposed on a satellite image is displayed., Event/Category/Experimental stimulus, (Item/Object/Vehicle/Aircraft/Airplane, Participant/Effect/Cognitive/Target, Sensory presentation/Visual/Rendering type/Screen/2D), (Item/Natural scene/Arial/Satellite, Sensory presentation/Visual/Rendering type/Screen/2D)",
            "Button": "..."
        }
    }
}
```
Here `mycodes` is an arbitrary column in a `events.tsv` file. The example shows how `mycodes` is described in the accompanying `events.json` file. Notably, see how `mycodes` is described by a dictionary of key value pairs. These key value pairs are usually among the following:

- `Description`
- `Units`
- ... (see: https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#tabular-files )

However, when `HED` is also among the key value pairs named above, then you can assume that the respective column (`mycodes` in this example) contains HED tags. Note, that this "advanced use" of HED tags is not yet implemented very well. It's described in the [HED appendix](https://bids-specification.readthedocs.io/en/stable/99-appendices/03-hed.html), but there is no mention of  `HED` as an admissible key for a key value pair in `events.json` in the [tabular-sidecar-file documentation](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#tabular-files). I also doubt that it is as of yet implemented in the bids-validator (I'll look for it later).

# TL;DR:

I suggest that for now you go with:

1. you can expect the `HED` column to have HED tags (that's the easiest)

Everything else is advanced usage and it's better to cover the basics first.

**Comment by sappelhoff** [comment]
indeed, HED is not yet accepted as the key for a `key-value-pair` in `events.json` files, although it *should* be as per the spec (see my earlier post).

Evidence: https://github.com/bids-standard/bids-validator/blob/master/bids-validator/validators/json/schemas/data_dictionary.json

This is an easy fix to add though

**Comment by sappelhoff** [comment]
regarding your second question:

>  More generally, how do I map any optional column header (such as HED) to a subscript in an array formed by splitting a row by tabs?

What do you mean by `a subscript in an array formed by splitting a row by tabs`? can you talk in terms of input and output? For an input `events.tsv` like the following:

```Text
duration    mycodes
1           a123
2           b234
1           c453
```

Do you want an output like this:

```js
['a123', 'b234', 'c453'] 
```

where each row is a string and rows are separated by commas?

**Comment by VisLab** [comment]
To reiterate the issue:
Somewhere around line 155 in validators/events.js we want to add the calls to validate the entries in the HED column.  However, at that point in events.js the column headers (HED specifically, but later we will want others) is not available.  We need this information to proceed.  We don't need to know how to index things, we just need the positions of the column headers.

Could someone please add this to the master branch so that we can proceed.  It will be generally necessary for a number of other validation steps downstream and as noted above, should have been there.  Thanks

**Comment by happy5214** [comment]
To clarify, the version of the file we're referring to is our development version (https://github.com/VisLab/bids-validator/blob/hed-integration/validators/events.js).

**Comment by sappelhoff** [comment]
Mmh I see, perhaps we can ask somebody for help. @nellh - would you have an idea?

**Comment by chrisgorgo** [comment]
It would be nice, but not crucial to have a dedicated function to parse TSV headers. It's a simple operation done by other validator procedures. Perhaps looking at the code for validating values of the `stim_file` column would help https://github.com/bids-standard/bids-validator/blob/b2f0dd3e6a37eb6caa2df0e0b16f535942636536/bids-validator/validators/tsv/tsv.js#L153

**Comment by happy5214** [comment]
Line 30 of that file creates a list of headers in a single line. I copied that over.

The issue on my end was the lack of headers on the existing test cases, which made me think the headers were stripped. I have working tests for the HED column using a header row, so I think this is a non-issue.

**Comment by chrisgorgo** [comment]
I'm glad this is resolved - I'm closing this for now. There is another old issue for consolidating TSV parsing #623.
